### PR TITLE
Constant-time shift-by-secret (ct_shl) + CT next_power_of_two for HeaplessBigInt

### PR DESCRIPTION
**Pull request by kaidokert**
Closes the heapless half of #171 — reinstates the Ct `NextPowerOfTwo` that was dropped in #170, now genuinely constant-time.

### `ct_shl` — barrel shifter (shift.rs)
A constant-time left shift by a **secret** amount. The inherent `Shl<usize>` branches and cache-indexes on `word_shift`/`bit_shift`, so shifting by a secret leaks it (that was #170's P1). The barrel shifter never lets the secret drive control flow:
- each stage `k` shifts by the **public** constant `2^k` (so `Shl`'s internal branches depend only on public `k`),
- and applies-or-not via a masked `ct_select` on bit `k` of the amount,
- over-width amounts mask to zero.

The only operations that touch the secret are `amount >> k & 1` and a `u32 >=` — both branchless. `O(width · log width)`, result width `self.len`.

### CT `next_power_of_two` (power_of_two.rs)
`next`/`wrapping_next_power_of_two` route their `1 << bits` (where `bits = width_bits - leading_zeros`, secret-derived) through `ct_shl` instead of `<<`, and combine the overflow flag with `&` (not short-circuiting `&&`). `leading_zeros`/`is_zero` already take the CT path on the `Ct` carrier. `checked_next_power_of_two` stays branchful (its `Option` reveals the overflow bit) — not CT, same documented caveat as FixedUInt; Ct callers use `next`/`wrapping`.

### Verification
CT is argued **by construction** (every secret-touching op enumerated above is branchless) and **correctness-tested**: `ct_shl` matches `<<` for all amounts incl. over-width; `next_power_of_two` matches std across normal / saturate / wrap / `None` cases.

**Empirical `ct-verify` coverage is a known gap, not addressed here:** the fixture harness is FixedUInt-only, and its asm-grep gate assumes straight-line branchless code — incompatible with a barrel-shifter's (public-bound) loop. Extending it to heapless + teaching the gate about public loops is separate infra work, and the **FixedUInt twin** (its Ct `next_power_of_two` has the same latent leak) is still open — both remain tracked in #171.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Add a constant-time left shift primitive for HeaplessBigInt and use it to implement a constant-time NextPowerOfTwo for the Ct personality while preserving existing Nct behavior.

New Features:
- Introduce ct_shl, a barrel-shift based constant-time left shift by a secret amount for HeaplessBigInt.
- Implement Ct-specific NextPowerOfTwo for HeaplessBigInt that uses ct_shl to avoid leaking the secret-derived shift amount.

Enhancements:
- Refactor HeaplessBigInt next_power_of_two logic into a shared checked_next_pow2 helper reusable across personalities.
- Expose max_at_len from heapless arithmetic as a reusable helper for computing width-dependent maximum values.

Tests:
- Add tests verifying ct_shl matches the standard left shift for all relevant shift amounts, including over-width.
- Add tests ensuring Ct next_power_of_two semantics match the standard behavior for non-overflow cases and handle zero and overflow (saturating, wrapping, checked) correctly.

### CODE REVIEW

File: src/heapless/shift.rs [hunk 1]
```diff
@@ -17,12 +17,48 @@
...
+    T: MachineWord + subtle::ConditionallySelectable,
+{
+    let width = value.len();
+    let word_bits = core::mem::size_of::<T>() as u32 * 8;
+    let width_bits = width as u32 * word_bits;
+    let mut result = value;
+    // Public loop bound: stages cover bit positions `0..ceil(log2(width_bits))`.
+    let mut k = 0u32;
+    while (1u32 << k) < width_bits {
+        let shifted = result << (1usize << k); // fixed, public shift by 2^k
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/shift.rs]: Fixed in c037a70 — the stage amount is now computed in u32 (1u32 << k, which stays < width_bits < 2^32 for the full u16 len range) and shifted via Shl<u32>, so k >= 16 no longer overflows on a 16-bit-usize target. The deeper residual (shift amounts > usize::MAX for >64K-bit values on 16-bit) is the same crate-wide usize limitation, and Shl<u32>'s over-width guard keeps it from truncating into a wrong small shift.
